server/queues/scoring/getTwoGames.coffee

```diff
@@ -95,46 +128,63 @@ findNextLeagueOpponent = (session, queryParams, callback) ->
+    delete query.randomSimulationIndex  # Just find the highest-indexed session, if our randomSimulationIndex was lower than the lowest one.
+    LevelSession.findOne(query).sort(sort).select(sessionSelectionString).lean().exec (err, session) ->
       return callback err if err
       callback null, session
 
-findRecentRandomSession = (queryParams, callback) ->
+findRecentRandomSession = (query, callback) ->
   # We pick a random submitDate between the first submit date for the level and now, then do a $lt fetch to find a session to simulate.
   # We bias it towards recently submitted sessions.
-  findEarliestSubmission queryParams, (err, startDate) ->
+  findEarliestSubmission query, (err, startDate) ->
     return callback err, null unless startDate
     now = new Date()
     interval = now - startDate
     cutoff = new Date now - Math.pow(Math.random(), 4) * interval
-    queryParams.submitDate = $gte: startDate, $lt: cutoff
-    LevelSession.findOne(queryParams).sort(submitDate: -1).select(sessionSelectionString).lean().exec (err, session) ->
+    query.submitDate = $gte: startDate, $lt: cutoff
+    LevelSession.findOne(query).sort(submitDate: -1).select(sessionSelectionString).lean().exec (err, session) ->
       return callback err if err
       callback null, session
 
 earliestSubmissionCache = {}
-findEarliestSubmission = (queryParams, callback) ->
-  cacheKey = JSON.stringify queryParams
+findEarliestSubmission = (query, callback) ->
+  cacheKey = JSON.stringify query
   return callback null, cached if cached = earliestSubmissionCache[cacheKey]
-  LevelSession.findOne(queryParams).sort(submitDate: 1).lean().exec (err, earliest) ->
+  LevelSession.findOne(query).sort(submitDate: 1).lean().exec (err, earliest) ->
     return callback err if err
     result = earliestSubmissionCache[cacheKey] = earliest?.submitDate
     callback null, result
+
+findComparableSession = (options, callback) ->
+  # We pick a random session kind of close in score to the target session
+  scores = options.scores
+  sessionRank = _.sortedIndex scores, options.session.totalScore, (index) -> -index  # Descending order
+  range = Math.max 100, Math.round(scores.length / 500)
+  maxScoreIndex = Math.max 0, sessionRank - range
+  minScoreIndex = Math.min scores.length - 1, sessionRank + range
+  possibleScores = scores.slice(maxScoreIndex, minScoreIndex)
+  targetScore = _.sample possibleScores
+  #console.log 'trying to find something with score close to', options.session.totalScore, 'of', scores.slice(0, 5), '... where we are ranked', sessionRank, ' and found', targetScore, 'between', maxScoreIndex, scores[maxScoreIndex], 'and', minScoreIndex, scores[minScoreIndex]
+  options.query.totalScore = $lte: targetScore
+  sort = totalScore: -1
+  LevelSession.findOne(options.query).sort(sort).select(sessionSelectionString).lean().exec (err, session) ->
+    return callback err if err
+    callback null, session
```

Comment:
This file in general could really use an overhaul. It would benefit from testing, using generators and avoiding editing function parameters. But it was like that before these changes and now after, so these changes don't make things worse. As is, I can't really speak to whether it works or not, though, unless we sit down and walk through it all. Given limited time, I say make sure it's thoroughly tested as much as you can before deploying it.